lib/modules/manager/npm/post-update/pnpm.ts

```diff
@@ -70,7 +76,19 @@ export async function generateLockFile(
       extraEnv.NPM_EMAIL = env.NPM_EMAIL;
     }
 
-    let args = '--recursive --lockfile-only';
+    const pnpmWorkspaceFilePath = getSiblingFileName(
+      lockFileName,
+      'pnpm-workspace.yaml',
+    );
+
+    let args = '--lockfile-only';
+    if (
+      pnpmToolConstraint.constraint &&
+      !semver.intersects(pnpmToolConstraint.constraint, '>=9') &&
+      (await localPathExists(pnpmWorkspaceFilePath))
+    ) {
+      args += ' --recursive';
```

Comment:
Please add a code comment describing why this is, for future maintainers/contributors